### PR TITLE
port to mac os x and g++-5

### PR DESCRIPTION
**Pull request by cc10512**
@antoninbas can you please review? If needed, I can go through these changes with you tomorrow.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -110,7 +110,8 @@ AC_CHECK_HEADERS([algorithm array cassert cmath queue \
 cstdio string sys/stat.h sys/types.h ctime tuple unistd.h unordered_map \
 utility vector], [], [AC_MSG_ERROR([Missing header file])])
 
-# Check for libjudy, libgmp, libnanomsg, libpcap
+# Check for pthread, libjudy, libgmp, libnanomsg, libpcap
+AC_CHECK_LIB([pthread], [pthread_create], [], [AC_MSG_ERROR([Missing pthread library])])
```
Comment (antoninbas) [inline configure.ac]: I was about to use the M4 `AX_PTHREAD` macro to check for pthread support. This macro checks for `libpthread` and handles the `-pthread` flag appropriately (with a special case for clang).

Comment (antoninbas) [inline configure.ac]: See http://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_pthread.m4

File: targets/simple_switch/tests/Makefile.am [hunk 2]
```diff
@@ -5,7 +5,7 @@ AM_CPPFLAGS += \
 -I$(srcdir)/.. \
 -I$(srcdir)/ \
 -DTESTDATADIR=\"$(srcdir)/testdata\"
-AM_CXXFLAGS = -pthread
+# AM_CXXFLAGS = -pthread
```
Comment (antoninbas) [inline targets/simple_switch/tests/Makefile.am]: just remove it

File: tests/Makefile.am [hunk 3]
```diff
@@ -12,7 +12,7 @@ libtestutils_la_LIBADD = $(top_builddir)/src/bm_sim/libbmsim.la
 AM_CPPFLAGS += \
 -isystem $(top_srcdir)/third_party/gtest/include \
 -DTESTDATADIR=\"$(srcdir)/testdata\"
-AM_CXXFLAGS = -pthread
+# AM_CXXFLAGS = -pthread
```
Comment (antoninbas) [inline tests/Makefile.am]: just remove it

File: tests/stress_tests/Makefile.am [hunk 4]
```diff
@@ -1,14 +1,14 @@
 AM_CPPFLAGS += \
 -isystem $(top_srcdir)/third_party \
 -DTESTDATADIR=\"$(abs_srcdir)/testdata\"
-AM_CXXFLAGS = -pthread
+# AM_CXXFLAGS = -pthread
```
Comment (antoninbas) [inline tests/stress_tests/Makefile.am]: remove

File: tests/test_calculations.cpp [hunk 5]
```diff
@@ -339,7 +342,7 @@ TEST(HashTest, Crc32) {
 }
 
 TEST(HashTest, Crc16Custom) {
-  typedef CustomCrcMgr<uint16_t>::crc_config_t crc16_config_t;
+  // typedef CustomCrcMgr<uint16_t>::crc_config_t crc16_config_t;
```
Comment (antoninbas) [inline tests/test_calculations.cpp]: just remove it if unused, no need to comment it out

File: tests/test_calculations.cpp [hunk 6]
```diff
@@ -384,7 +387,7 @@ TEST(HashTest, Crc16Custom) {
 }
 
 TEST(HashTest, Crc32Custom) {
-  typedef CustomCrcMgr<uint32_t>::crc_config_t crc32_config_t;
+  // typedef CustomCrcMgr<uint32_t>::crc_config_t crc32_config_t;
```
Comment (antoninbas) [inline tests/test_calculations.cpp]: same here, you can just remove the line

File: tests/test_checksums.cpp [hunk 7]
```diff
@@ -40,6 +40,7 @@ static const unsigned char raw_tcp_pkt[66] = {
   0x21, 0x9b                                      /* !. */
 };
 
+#if ENABLE_ME_WHEN_USED
```
Comment (antoninbas) [inline tests/test_checksums.cpp]: if it is unused, remove this block of code. I will always be able to find it later if I need to.

File: tests/test_devmgr.cpp [hunk 8]
```diff
@@ -39,7 +39,21 @@ class TestDevMgrImp : public DevMgrIface {
...
+    //             ^
+    // ../../tests/test_devmgr.cpp:42:17: note: remove std::move call here
+    // p_monitor = std::move(PortMonitorIface::make_active(0));
+    //             ^~~~~~~~~~                                ~
+	// Disabling the warning seems more appropriate than removing the move
+#pragma GCC diagnostic push
+#if defined(__clang__) && (__clang_major__ > 7)
+#pragma GCC diagnostic ignored "-Wpessimizing-move"
+#endif // __clang__
+	p_monitor = std::move(PortMonitorIface::make_active(0));
```
Comment (antoninbas) [inline tests/test_devmgr.cpp]: the compiler is right; you should just remove the `move` here. Leaving the `move` prevents the compiler optimization and makes the code less readable.

File: tests/test_fields.cpp [hunk 9]
```diff
@@ -196,7 +196,7 @@ class TwoCompV {
 
  private:
   ByteContainer bits_{};
-  int nbits_{0};
+  // int nbits_{0};
```
Comment (antoninbas) [inline tests/test_fields.cpp]: remove if not used (and the initialization code above as well)

File: tests/test_header_stacks.cpp [hunk 10]
```diff
@@ -164,7 +164,7 @@ TEST_F(HeaderStackTest, PushFrontNum) {
   Header &h2 = phv->get_header(testHeader_2);
 
   Field &f0_0 = h0.get_field(0);
-  Field &f1_0 = h1.get_field(0);
+//  Field &f1_0 = h1.get_field(0);
```
Comment (antoninbas) [inline tests/test_header_stacks.cpp]: remove

File: tests/test_header_stacks.cpp [hunk 11]
```diff
@@ -199,7 +199,7 @@ TEST_F(HeaderStackTest, PopFront) {
 
   Field &f0_0 = h0.get_field(0);
   Field &f1_0 = h1.get_field(0);
-  Field &f2_0 = h2.get_field(0);
+//  Field &f2_0 = h2.get_field(0);
```
Comment (antoninbas) [inline tests/test_header_stacks.cpp]: remove

File: tests/test_p4objects.cpp [hunk 12]
```diff
@@ -326,7 +326,7 @@ TEST(P4Objects, HeaderStackArith) {
   ASSERT_NO_THROW(h0_f1.get_int());
 
   if (!WITH_VALGRIND) {
-    ASSERT_DEATH(h0_f2.get_int(), "Assertion .* failed");
-    ASSERT_DEATH(h1_f2.get_int(), "Assertion .* failed");
```
Comment (antoninbas) [inline tests/test_p4objects.cpp]: not sure what is the point of this but if it works it is fine

File: tests/test_tables.cpp [hunk 13]
```diff
@@ -1368,8 +1356,8 @@ TEST_F(TableIndirectWS, EntryWS) {
   MatchErrorCode rc;
   grp_hdl_t grp;
   mbr_hdl_t mbr_1;
-  mbr_hdl_t mbr_bad = 999u;
-  mbr_hdl_t grp_bad = 999u;
+  // mbr_hdl_t mbr_bad = 999u;
```
Comment (antoninbas) [inline tests/test_tables.cpp]: remove if unused

File: tests/test_tables.cpp [hunk 14]
```diff
@@ -2776,7 +2764,7 @@ TEST_F(TableRangeMatch, TwoRanges) {
   ActionData action_data;
   std::vector<MatchKeyParam> match_key;
   unsigned int start_1 = 0xab, start_2 = 0x8888;
-  unsigned int end_1 = 0x041d, end_2 = 0x8889;
+  // unsigned int end_1 = 0x041d, end_2 = 0x8889;
```
Comment (antoninbas) [inline tests/test_tables.cpp]: remove

File: tests/test_p4objects.cpp (hunk 12 above)
Comment (cc10512) [inline tests/test_p4objects.cpp]: clang on the mac prints the assertion as "Assertion failed: (arith) ...", which fails the regex you had before which implies two spaces.

**Comment by cc10512** [comment]
@antoninbas I addressed your requests. Please take another look.

File: tests/primitives.cpp [hunk 15]
```diff
@@ -49,6 +49,8 @@ REGISTER_PRIMITIVE(add_to_field);
 class generate_digest : public ActionPrimitive<const Data &, const Data &> {
   void operator ()(const Data &receiver, const Data &learn_id) {
     // stub only
+		(void)receiver;
```
Comment (antoninbas) [inline tests/primitives.cpp]: indentation is wrong, and you are using tabs?

File: tests/test_devmgr.cpp [hunk 16]
```diff
@@ -39,7 +39,7 @@ class TestDevMgrImp : public DevMgrIface {
 public:
   TestDevMgrImp() {
     // 0 is device_id
-    p_monitor = std::move(PortMonitorIface::make_active(0));
+		p_monitor = PortMonitorIface::make_active(0);
```
Comment (antoninbas) [inline tests/test_devmgr.cpp]: indentation and tabs?

**Comment by cc10512** [comment]
@antoninbas sorry about that. fixed the indentation and the tabs and in the process a number of empty lines that had tabs and whitespace have also snuck in. Hope it's ok.

**Comment by antoninbas** [comment]
@cc10512 As far as I can tell you forgot at least utils.h and primitives.cpp

**Comment by cc10512** [comment]
@antoninbas: @cc10512 ducks.

**Pull request merged**